ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/fullsync/FullDownloaderFactory.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.eth.sync.fullsync;
+
+import tech.pegasys.pantheon.ethereum.ProtocolContext;
+import tech.pegasys.pantheon.ethereum.eth.manager.EthContext;
+import tech.pegasys.pantheon.ethereum.eth.sync.ChainDownloader;
+import tech.pegasys.pantheon.ethereum.eth.sync.SynchronizerConfiguration;
+import tech.pegasys.pantheon.ethereum.eth.sync.TrailingPeerRequirements;
+import tech.pegasys.pantheon.ethereum.eth.sync.state.SyncState;
+import tech.pegasys.pantheon.ethereum.mainnet.ProtocolSchedule;
+import tech.pegasys.pantheon.metrics.MetricsSystem;
+
+public class FullDownloaderFactory<C> {
+
+  private final ChainDownloader chainDownloader;
+  private final SynchronizerConfiguration syncConfig;
+  private final ProtocolContext<C> protocolContext;
+  private final SyncState syncState;
+
+  public FullDownloaderFactory(
+      final SynchronizerConfiguration syncConfig,
+      final ProtocolSchedule<C> protocolSchedule,
+      final ProtocolContext<C> protocolContext,
+      final EthContext ethContext,
+      final SyncState syncState,
+      final MetricsSystem metricsSystem) {
+    this.syncConfig = syncConfig;
+    this.protocolContext = protocolContext;
+    this.syncState = syncState;
+
+    this.chainDownloader =
+        FullSyncChainDownloader.create(
+            syncConfig, protocolSchedule, protocolContext, ethContext, syncState, metricsSystem);
+  }
+
+  public void start() {
+    chainDownloader.start();
```

Comment:
To follow-through on the factory change, should this class now `create()` a `Downloader` that has a `start()` method?  Just seems weird to call `Factory.start()`.

Reply:
We can't really because there are additional methods the "factories" need to expose - trailing peers for full sync and deleting the state for fast.

Reply:
Maybe these are `FastDownloadManager` and `FullDownloadManager`?  Manager is a bit of a weasel word but I'm not seeing a better layout that would give clearer responsibilities at the moment.

Reply:
I think you could get away with a `Downloader` that just exposes `start` and `calculateTrailingPeerRequirements`, then does any cleanup before the `start` future completes.  But feel free to just rename if you think that makes more sense.

Reply:
I'm not sure we've got this structure quite right yet but I think I've found a reasonable landing point.  Fast sync has a `FastSyncDownloadFactory` which returns a `FastSyncDownload` - full sync is much simpler so doesn't need the factory and just creates a `FullSyncDownload` directly.

End result is that `DefaultSynchronizer` winds up with a `Optional<FastSyncDownload>` and a `FullSyncDownload` both of which have `start` and `calculateTrailingPeerRequirements`.

There's more tidy-up that could be done in this area but I'm hesitant to get too distracted.